esmvalcore/_recipe.py

```diff
@@ -330,77 +333,64 @@ def _add_fxvar_keys(fx_var_dict, variable):
+            f"with parent variable '{variable}' does not have "
+            f"a '{var_project}' project in config-developer.")
+    project_tables = CMOR_TABLES[var_project].tables
 
     # force only the mip declared by user
-    if 'mip' in fx_variable:
-        fx_mips = [fx_variable['mip']]
+    found_mip = False
+    if not fx_info['mip']:
+        for mip in project_tables:
+            fx_cmor = project_tables[mip].get(fx_info['short_name'])
+            if fx_cmor:
+                found_mip = True
+                fx_info['mip'] = mip
+                fx_info = _add_fxvar_keys(fx_info, variable)
+                logger.debug(
+                    "For fx variable '%s', found table '%s'",
+                    fx_info['short_name'], mip)
+                fx_files = _get_input_files(fx_info, config_user)[0]
+                if fx_files:
+                    logger.debug(
+                        "Found fx variables '%s':\n%s",
+                        fx_info['short_name'],
+                        pformat(fx_files))
+                    break
     else:
-        # Get all fx-related mips (original var mip,
-        # 'fx' and extend from cmor tables)
-        fx_mips = [variable['mip']]
-        fx_mips.extend(mip for mip in cmor_table.tables if 'fx' in mip)
-
-    # Search all mips for available variables
-    # priority goes to user specified mip if available
-    searched_mips = []
-    fx_files = []
-    for fx_mip in fx_mips:
-        fx_cmor_variable = cmor_table.get_variable(fx_mip, fx_varname)
-        if fx_cmor_variable is not None:
-            fx_var_dict = dict(fx_variable)
-            searched_mips.append(fx_mip)
-            fx_var_dict['mip'] = fx_mip
-            fx_var_dict = _add_fxvar_keys(fx_var_dict, var)
-            valid_fx_vars.append(fx_var_dict)
-            logger.debug("For fx variable '%s', found table '%s'", fx_varname,
-                         fx_mip)
-            fx_files = _get_input_files(fx_var_dict, config_user)[0]
-
-            # If files found, return them
-            if fx_files:
-                logger.debug("Found fx variables '%s':\n%s", fx_varname,
-                             pformat(fx_files))
-                break
+        fx_cmor = project_tables[fx_info['mip']].get(fx_info['short_name'])
+        if fx_cmor:
+            found_mip = True
+            fx_info = _add_fxvar_keys(fx_info, variable)
+            fx_files = _get_input_files(fx_info, config_user)[0]
 
     # If fx variable was not found in any table, raise exception
-    if not searched_mips:
+    if not found_mip:
         raise RecipeError(
-            f"Requested fx variable '{fx_varname}' not available in "
-            f"any 'fx'-related CMOR table ({fx_mips}) for '{var_project}'")
+            f"Requested fx variable '{fx_info['short_name']}' "
+            f"not available in any CMOR table for '{var_project}'")
 
     # flag a warning
     if not fx_files:
```

Comment:
note that this will fail with `fx_files` variable not defined if `fx_cmor` is False, so prob move this up in the loop

Reply:
But a False fx_cmor raises a RecipeError, so the tool would stop before getting here no?

Reply:
Sorry what I meant is that if `fx_cmor` is `None`, then `found_mip` will stay `False` (as in line 325). And that will raise the error before getting to line 384.  Anyway I don't think I changed the order of reporting that there is already in master.

Reply:
OK cool, I trust you on that :grin: